test/base/src/androidTest/kotlin/io/realm/test/shared/RealmTests.kt

```diff
@@ -478,6 +482,114 @@ class RealmTests {
+        val testDirPath = testDir.toPath()
+        assertTrue(fileSystem.exists(testDirPath))
+
+        val configuration = RealmConfiguration.Builder(schema = setOf(Parent::class, Child::class))
+            .path("$testDir/default.realm")
+            .build()
+        val bgThreadReadyChannel = Channel<Unit>(1)
+        val readyToCloseChannel = Channel<Unit>(1)
+        val closedChannel = Channel<Unit>(1)
+
+        runBlocking {
+            // Creates another Realm to ensure the log files are generated.
+            val testRealm = Realm.open(configuration)
+
+            val deferred = async {
+                val anotherRealm = Realm.open(configuration)
+
+                bgThreadReadyChannel.send(Unit)
+                readyToCloseChannel.receive()
+
+                anotherRealm.close()
+
+                closedChannel.send(Unit)
+            }
+
+            testRealm.writeBlocking {
+                copyToRealm(Child())
+            }
+
+            // Waits for bg thread's opening the same Realm.
+            bgThreadReadyChannel.receive()
+
+            // A core upgrade might change the location of the files
+            val testDirRenamedPath = PlatformUtils.createTempDir("test_dir_renamed").toPath()
+            assertTrue(fileSystem.exists(testDirRenamedPath))
+            fileSystem.atomicMove(testDirPath, testDirRenamedPath)
+            val testDirRenamedPathList = fileSystem.list(testDirRenamedPath)
+            assertEquals(4, testDirRenamedPathList.size) // db file, .lock, .management, .note
+
+            readyToCloseChannel.send(Unit)
+
+            testRealm.close()
+
+            closedChannel.receive()
+
+            // Now we get log files back!
+            fileSystem.atomicMove(testDirRenamedPath, testDirPath)
+
+            assertTrue(Realm.deleteRealm(configuration))
+
+            val testDirPathList = fileSystem.list(testDirPath)
+            // Lock file should never be deleted
+            assertEquals(1, testDirPathList.size) // only .lock file remains
+            assertTrue(fileSystem.exists("${configuration.path}.lock".toPath()))
+
+            deferred.cancel()
+            bgThreadReadyChannel.close()
+            readyToCloseChannel.close()
+            closedChannel.close()
+        }
+    }
+
+    @Test
+    fun deleteRealm_failures() {
+        // val tempDirA = PlatformUtils.createTempDir()
+        // val pathA = "$tempDirA/default.realm"
+        val configA = RealmConfiguration.Builder(schema = setOf(Parent::class, Child::class))
+            // .path(pathA)
+            .path("$tmpDir/default.realm")
+            .build()
+        val tempDirB = PlatformUtils.createTempDir()
+        val pathB = "$tempDirB/yetAnotherRealm.realm"
+        val configB = RealmConfiguration.Builder(schema = setOf(Parent::class, Child::class))
+            .path(pathB)
+            .build()
+
+        // This instance is already cached because of the setUp() method so this deletion should throw.
+        try {
```

Comment:
There is `assertFailsWith<IllegalStateException>` which you can use instead of the try-catch